### PR TITLE
Fire - Pauline

### PR DESCRIPTION
**Pull request by ghostfruitleaf**
# Task List
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe in your own words what the Model is doing in Rails | The Model is the data that requests to the application are meant to operate on. More specifically, models dictate how the data inside the application are organized through an object-oriented lens.
Describe in your own words what the Controller is doing in Rails | The controller processes requests between the user and its models through various functions defined regarding what it can and can't do, preferably using RESTful routes in creating, updating, reading, and deleting model data, and returning the result to the user through view.
Describe in your own words what the View is doing in Rails | The view is the controller's way of presenting the result of a request made from a user -- whether this is displaying a form, deleting an entry, creating an entry, or editing data in a certain format that the application wants the user to view the application's contents as. 
Describe an edge-case controller test you wrote | One edge controller test I wrote was used to test the destroy method. This edge case described a situation where a user attempts to delete a task that doesn't exist -- for instance, if the user has the same page open in two windows, and deletes the task in the first window, an attempt to delete the same task in the second window (since at this point the page is not responsive) would result in a 404 error as called by `head :not_found` in the TaskController class. The edge case had to ensure that the controller responded with `:not_found` to trigger the 404 page and that no objects were accidentally deleted, meaning `Task.count `wouldn't change in value.
What is the purpose of using strong params? (i.e. the params method in the controller) | Strong params, namely within a method, add an extra layer of protection to what information from the database is read, modified, created, or destroyed. In rails, controllers cannot directly pass parameters to the the ActiveRecord from, for instance, a form used to collect data for those parameters in order to protect sensitive fields within the database. By defining strong params that explicitly describe what fields can be utilized, Rails will in turn allow through those strong params the attributes that controller can use for its methods.
How are Rails migrations related to Rails models? | Migrations can be said to be an implementation of how to store data as described by the models, as well as allow the models' schemas to be modified and track those modifications. Very much like the internal code to describe an Abstract Data Type, Migrations can modify the data as needed, likely to best reflect the purpose of the Model in the context of the application.
Describe one area of Rails that you are still unclear on | I think I still need to take time and completely understand RESTful routes, as well as all the different things you can do in postgresql and how to migrate changes to the database. I'm finding I'm having a hard time understanding how Minitest is testing Rails, and had to roll back a couple functionalities because I couldn't figure out how to test them and find information on the internet that wasn't too esoteric for me to understand.

### CODE REVIEW

File: app/controllers/tasks_controller.rb [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+      render :edit
+      return
+    end
+  end
+
+  # DESTROY
+  def destroy
+    task_id = params[:id]
+    @task = Task.find_by(id: task_id)
+    # NOTE: confirmation page is handled by index and show pages as a dialog box.
```
Comment (tildeee) [inline app/controllers/tasks_controller.rb:63]: Good comment!

File: test/controllers/tasks_controller_test.rb [hunk 2]
```diff
@@ -83,42 +79,193 @@
       must_respond_with :redirect
       must_redirect_to task_path(new_task.id)
     end
+
+    it "will render :edit if attempts to enter an empty task are made" do
+      # Arrange
+      empty_task = {task: {name:nil, description:nil}}
```
Comment (tildeee) [inline test/controllers/tasks_controller_test.rb:85]: well-written tests! I think they read really well because your Arrange step always makes it clear how it's "invalid" if it is

File: test/controllers/tasks_controller_test.rb [hunk 3]
```diff
@@ -83,42 +79,193 @@
...
   end
   
-  # Complete these tests for Wave 4
+  # DONE - Complete these tests for Wave 4
   describe "destroy" do
-    # Your tests go here
-    
+    it "can destroy a model" do
+      # Arrange
+      delete_me = Task.new name: "Delete me", description: "Delete me"
```
Comment (tildeee) [inline test/controllers/tasks_controller_test.rb:184]: This syntax is correct, but it doesn't have the parentheses. To match the rest of your code style (which uses a lot of parens, namely around Task.new), it might be more consistent to have `Task.new(name: "", description:"")`

**Review by tildeee — COMMENTED** [review]
# Task List




## Major Learning Goals/Code Review





| Criteria                                           | yes/no, and optionally any details/lines of code to reference |
| -------------------------------------------------- | ------------------------------------------------------------- |
| At least 6 commits with meaningful commit messages | ✔️                                                            |
| Routes follow RESTful conventions                  | ✔️                                                            |
| Uses named routes (like `_path`)                   | ✔️                                                            |
| Creates Models and migrations                      | ✔️                                                            |
| Creates styled views                               | ✔️                                                            |
| Handles errors like nonexistant tasks              | ✔️                                                            |
| Uses `form_with` to render forms in Rails          | ✔️                                                            |

## Functional Requirements/Manual Testing

| Functional Requirement                                                         | yes/no |
| ------------------------------------------------------------------------------ | ------ |
| Successfully handles index & show                                              | ✔️     |
| index & show tests pass                                                        | ✔️     |
| Successfully handles: New, Create                                              | ✔️     |
| New, Create tests pass                                                         | ✔️     |
| Successfully handles: Edit, Update                                             | ✔️     |
| Edit, Update tests pass with valid & invalid task ids                          | ✔️     |
| Successfully handles: Destroy, Task Complete                                   | ✔️     |
| Tests for Destroy & Task Complete include tests for valid and invalid task ids | ✔️     |

## Overall Feedback

| Overall Feedback                | Criteria                                                                                                                                                                                      | yes/no |
| ------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------ |
| Green (Meets/Exceeds Standards) | 5+ in Code Review && 6+ in Functional Requirements                                                                                                                                            | ✔️
| Yellow (Approaches Standards)   | 3+ in Code Review && 5+ in Functional Requirements, or the instructor judges that this project needs special attention                                                                        |
| Red (Not at Standard)           | 0-2 in Code Review or 0-4 in Functional Reqs, or assignment is breaking/doesn’t run with less than 5 minutes of debugging, or the instructor judges that this project needs special attention |

### Additional Feedback

Pauline! Great work on this project!!

Your project works and looks great-- the finished submission is fun. I like the layout you decided; you put thought into making the text BIG and noticeable! :D

Your code reads really well and looks exactly like how I expect it to look. Overall, I think your tests are particularly well-written and readable :D And I like your implementation for `toggle_complete`

Well done!

## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality              | Yes? |
| -------------------- | ---- |
| Perfect Indentation  | ✅    |
| Elegant/Clever       |     |
| Descriptive/Readable | ✅    |
| Concise              |    |
| Logical/Organized    | ✅    |
